gk/main.c

```diff
@@ -1212,51 +1660,1034 @@ setup_gk_lpm(struct gk_config *gk_conf, unsigned int socket_id)
 		goto free_lpm;
 	}
 
+	ret = init_fib_tbl(gk_conf);
+	if (ret < 0)
+		goto free_lpm6;
+
 	ret = 0;
 	goto out;
 
+free_lpm6:
+	destroy_ipv6_lpm(ltbl->lpm6);
+
 free_lpm:
 	destroy_ipv4_lpm(ltbl->lpm);
 
 out:
 	return ret;
 }
 
-static int
-gk_stage1(void *arg)
+/*
+ * Initialize a gateway FIB entry.
+ * @gw_addr the gateway address informaiton.
+ * @ip_prefix_addr the IP prefix,
+ * for which the gateway is responsible.
+ * @ip_prefix_len the prefix length.
+ */
+static struct gk_fib *
+init_gateway_fib(struct ipaddr *gw_addr, struct ipaddr *ip_prefix_addr,
```

Comment:
Adding more than 1K lines of code in `gk/main.c` to deal with the maintenance of the FIB table, what it not the main role of GK blocks, is not a good way to structure the code because it obfuscates what GK blocks really do and makes the code unmaintainable. Please move the FIB code to a new file. Suggestion: "gk/fib.c"